python/sdist/amici/sbml_utils.py

```diff
@@ -0,0 +1,558 @@
+        if replace_time:
+            expr = expr.subs(amici_time_symbol, sbml_time_symbol)
+    return MathMLSbmlPrinter(settings).doprint(expr, pretty=pretty)
+
+
+def sbml_math_ast(expr, **kwargs) -> libsbml.ASTNode:
+    """Convert a SymPy expression to SBML math AST.
+
+    :param expr:
+        expression to be converted (will be sympified).
+
+    :param kwargs:
+        extra options for MathML conversion.
+    """
+    mathml = sbml_mathml(expr, **kwargs)
+    ast = libsbml.readMathMLFromString(mathml)
+    if ast is None:
+        raise SbmlMathError(
+            f'error while converting the following expression to SBML AST.\n'
+            f'expression:\n{expr}\n'
+            f'MathML:\n{pretty_xml(mathml)}'
+        )
+    return ast
+
+
+def set_sbml_math(obj: libsbml.SBase, expr, **kwargs) -> None:
+    """Set the math attribute of a SBML node using a SymPy expression.
+
+    :param obj:
+        SBML node supporting `setMath` method.
+
+    :param expr:
+        expression to which the math attribute of `obj` should be se to
+        (will be sympified).
+
+    :param kwargs:
+        extra options for MathML conversion.
+    """
+    mathml = sbml_math_ast(expr, **kwargs)
+    if obj.setMath(mathml) != libsbml.LIBSBML_OPERATION_SUCCESS:
+        raise SbmlMathError(
+            f'Could not set math attribute of SBML object {obj}\n'
+            f'expression:\n{expr}\n'
+            f'MathML:\n{pretty_xml(mathml)}'
+        )
+
+
+###############################################################################
+# MathML to Sympy conversion
+
+
+def mathml2sympy(
+    mathml: str,
+    *,
+    evaluate: bool = False,
+    locals: Optional[Dict[str, Any]] = None,
+    expression_type: str = 'mathml2sympy',
+) -> sp.Basic:
+    ast = libsbml.readMathMLFromString(mathml)
+    if ast is None:
+        raise ValueError(
+            f'libSBML could not parse MathML string:\n{pretty_xml(mathml)}'
+        )
+
+    formula = _parse_logical_operators(libsbml.formulaToL3String(ast))
+
+    if evaluate:
+        expr = sp.sympify(formula, locals=locals)
+    else:
+        with sp.core.parameters.evaluate(False):
+            expr = sp.sympify(formula, locals=locals)
+
+    expr = _parse_special_functions(expr)
+
+    if expression_type is not None:
+        _check_unsupported_functions(expr, expression_type)
+
+    return expr
+
+
+def _parse_logical_operators(math_str: Union[str, float, None]
+                             ) -> Union[str, float, None]:
+    """
+    Parses a math string in order to replace logical operators by a form
+    parsable for sympy
+
+    :param math_str:
+        str with mathematical expression
+    :param math_str:
+        parsed math_str
+    """
+    if not isinstance(math_str, str):
+        return math_str
+
+    if ' xor(' in math_str or ' Xor(' in math_str:
+        raise SBMLException('Xor is currently not supported as logical '
+                            'operation.')
```

Comment:
There is `sympy.Xor`, but I think this function was just moved here. So okay to leave.